Ability to set a country, both in server and user settings, to have the correct streaming provider (but doesn't affect language)

stignarnia opened the pull request:
I felt like this feature was needed, and I also found issues #452 and #117 mentioning it!

I'm very inexperienced and this is one of my first PRs in general, so I know what I did works but it might be stupid, have a wrong style or who knows what, so I'm sorry in advance and feel free to reject this or make any necessary changes.

### Changes made

#### You can choose the default country new users will be onboarded in
![immagine](https://github.com/sbondCo/Watcharr/assets/80171209/d04b82cf-ba12-45bd-9164-3fec1261696a)

#### Users can choose the country for the streaming providers
![immagine](https://github.com/sbondCo/Watcharr/assets/80171209/d5a4f3cd-fd74-4ab9-90ce-0934fe34f4ac)

Commit:
Country support for watch providers (#1, not functional)

Commit:
frontend dropdowns now actually read and write the correct setting

Commit:
user country setting now respects the server side default when creating a new user

Commit:
watch providers correctly show based on country, but changing the coutnry requires restarting the container to take effect

Commit:
flushing cache when updating user settings

Commit:
non functional, initial support for getting all the countries of which tmdb has streaming providers for to let users pick one

Commit:
users get shown a list of country codes to pick from and the selection gets saved and used, but the dropdown menu doesn't allow for scrolling

Commit:
limited dropdown height and auto scroll to the letter the user hits

Commit:
simplified auto scroll code

Commit:
the dropdown to choose countries now shows their full english name

IRHM committed:
server/user region improvements

IRHM commented:
Hey @stignarnia, I apologize for it taking so long before I could look at this PR. I really appreciate your great work!

I did make a few modifications (some I've had planned in my head for a while), hopefully that's ok! I thought it would be a little easier than asking for a million changes from you.

So that you understand my reasoning, I'll add a note on each change I've made (hopefully they make sense and can help).

Thank you again!

IRHM commented on server/auth.go:
```diff
@@ -421,6 +425,7 @@ func loginJellyfin(user *User, db *gorm.DB) (AuthResponse, error) {
 			dbUser.ThirdPartyAuth = resp.AccessToken
 			dbUser.Username = resp.User.Name
 			dbUser.Type = JELLYFIN_USER
+			dbUser.Country = &Config.DEFAULT_COUNTRY
```
The `loginJellyfin` and `loginPlex` methods will create a user if they don't exist in our database, so added the default country for them too.

IRHM commented on server/middleware.go:
```diff
@@ -8,13 +8,24 @@ import (
 
 // Location middleware
 func WhereaboutsRequired() gin.HandlerFunc {
```
I changed this around to look for a `query` parameter on the request (I had planned to do this a little while ago #117).

The issue explains in more detail, but the tl;dr is it saves us making an extra database call and means we don't need to flush cache each time the user changes their settings.

IRHM commented on server/routes.go:
```diff
@@ -122,7 +122,7 @@ func (b *BaseRouter) addContentRoutes() {
 
 	// Get tv details (for tv page)
-	content.Use(WhereaboutsRequired()).GET("/tv/:id", cache.CachePage(b.ms, exp, func(c *gin.Context) {
+	content.GET("/tv/:id", WhereaboutsRequired(), cache.CachePage(b.ms, exp, func(c *gin.Context) {
```
Moved the usage of `WheraboutsRequired`. When adding it with `.Use`, the middleware will apply to all subsequently defined routes. 

Adding it as a handler before our main handler func means it will only apply to this route.

IRHM replied:
I just realised, this might have been me that added it with `.Use` 💀

IRHM committed:
DropDown: remove console log

used to test

IRHM commented on src/lib/DropDown.svelte:
```diff
@@ -11,15 +14,54 @@
+    const btns = ulElement?.querySelectorAll("button");
+    for (let i = 0; i < btns.length; i++) {
+      const btn = btns[i];
+      if (btn.textContent?.toLowerCase()?.startsWith(pressedLetter)) {
```
Main change, it'll scroll to the first button with text starting with the letterPressed.

Had to change so it would work with the new way the data is passed.

May have to change in the future, I think it'll only work well if the data in the dropdown is sorted alphabetically, which it isn't guaranteed to be (for now it is, with regions).

stignarnia replied:
I'm not sure it is actually, I think it's sorted alphabetically by county code, but we display county names

IRHM replied:
@stignarnia Oh that makes sense. I changed the way the data is provided to the dropdown, so guess I had to change this anyways (I've forgot now, but I believe there was no way to get the button by it's full value anymore after my change so I decided to just rely on the first letter).

stignarnia replied:
I can't say I understand what you did, but I tested it, and it works. For example, "VA" is correctly sorted among other countries starting with H as the English name is "Holy See".

IRHM commented on src/lib/DropDown.svelte:
```diff
@@ -11,15 +14,54 @@
   }
+
+  onMount(() => {
+    document.addEventListener("keypress", handleKeyPress);
```
Moved this to onMount so it's only added once.

IRHM commented on src/lib/DropDown.svelte:
```diff
@@ -11,15 +14,54 @@
+  onMount(() => {
+    document.addEventListener("keypress", handleKeyPress);
+
+    return () => {
```
Added this, so when the component is destroyed (eg user navigates away to different page) we remove the event listener.

IRHM commented on src/lib/RegionDropDown.svelte:
Created new RegionDropDown component so we could reuse the logic for getting regions.

IRHM committed:
DropDown: Add keypress event listener on main div

instead of document body, so only triggered when key pressed whilst focused on dropdown.

stignarnia commented:
Thank you so much for taking a look at it! Ofc changes are welcomed, they all make sense except there may be a problem with the auto scroll one. Thank you for merging as well!
